### PR TITLE
fix: consistently use ErrorDetail in StackTrace

### LINKED ISSUE #225: StackTrace frames are strings for unknown contracts
The `StackTrace` class is receiving `str` frames instead of `ErrorDetail`. This happens in:

- The ABI contract traces:
https://github.com/vyperlang/titanoboa/blob/1bf16f916b91aab299e293b4148be455c1674706/boa/contracts/abi/abi_contract.py#L273

- Unknown called contracts:
https://github.com/vyperlang/titanoboa/blob/1bf16f916b91aab299e293b4148be455c1674706/boa/contracts/base_evm_contract.py#L63

We should convert these to create proper `ErrorDetail` instances.

### PR DESCRIPTION
**Pull request by DanielSchiavini**
Fixes #225

### What I did
Converted the stack traces for ABI and unknown contracts to a list of `ErrorDetail` instead of `str`.

### How I did it
- Moved some Vyper contract handling to the base EVM contract
- Moved loose error functions to BoaError class method

### How to verify it
- Tests are included

### Description for the changelog
- StackTraces for ABI have expose more information

### Cute Animal Picture
image

### CODE REVIEW

File: boa/contracts/vyper/vyper_contract.py [hunk 1]
```diff
@@ -499,16 +397,18 @@ def __init__(
 
         # set external methods as class attributes:
         self._ctor = None
+        self._address = None
```
Comment (charles-cooper) [inline boa/contracts/vyper/vyper_contract.py:400]: is this change necessary?

File: boa/contracts/vyper/vyper_contract.py [hunk 2]
```diff
@@ -499,16 +397,18 @@ def __init__(
...
         if "__init__" in external_fns:
             self._ctor = VyperFunction(external_fns.pop("__init__"), self)
 
         if skip_initcode:
             if value:
                 raise Exception("nonzero value but initcode is being skipped")
-            addr = Address(override_address)
+            self._address = Address(override_address)
```
Comment (charles-cooper) [inline boa/contracts/vyper/vyper_contract.py:407]: is this change necessary?

File: boa/contracts/vyper/vyper_contract.py [hunk 3]
```diff
@@ -499,16 +397,18 @@ def __init__(
...
         else:
-            addr = self._run_init(*args, value=value, override_address=override_address)
-        self._address = addr
```
Comment (charles-cooper) [inline boa/contracts/vyper/vyper_contract.py:511]: not sure why this was removed, it makes the behavior at the join point clear

Comment (DanielSchiavini) [reply boa/contracts/vyper/vyper_contract.py:511]: I didn't realize the attribute was defined in the base class. Will revert

Btw this is why i lke ternaries in general

**Review by charles-cooper — COMMENTED** [review]
hmm overall i'm not sure this PR needs to go so far. ErrorDetail has vyper-specific info in it; doesn't seem that it needs to be moved into base_evm_contract.py

**Comment by DanielSchiavini** [comment]
I'll make https://github.com/vyperlang/titanoboa/pull/245 independent on this one and close #225. After the last commit the changes don't add much value.
